### PR TITLE
CANS-278: Worker can only see children in their county

### PR DESCRIPTION
**Pull request by edgibbs**
This filters out all the child records based on county using a Hibernate filter on Person. It completes part of the full functionality for the story. This PR was done test first, and to support this I had to add a fair amount of unit testing infrastructure.  Note, unit tests were primarily only added for new functionality in this PR. And lastly feel free to critque away, my Java is rusty, and many of these frameworks are realtively new to me.

https://osi-cwds.atlassian.net/browse/CANS-278
- [x] I used TDD to develop the feature
- [x] I have included unit tests
- [x] I have included new acceptance tests or modified existing ones
- [ ] I have included other tests

- [x] New feature (non-breaking change which adds functionality)
- [ ] Refactoring (No behavioral changes)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

- [x] My code follows the code style of this project.
- [x] I ran all my tests locally which were green.
- [ ] My code adds no new issues to Code Climate
- [x] I ran all the linters and static analyzers for the project (SonarQube, eslint, rubocop, etc).
- [x] My code is in a stable state ready to be deployable, but not necessarily complete.
- [x] I promise on my honor as a CWDS developer to ensure I don't break the build, to check linters, use best practices, to leave the code in better shape than I found it, and to have a good day.

### CODE REVIEW

File: src/test/java/gov/ca/cwds/cans/dao/CountyDaoTest.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+  @Test(expected = NullOrEmptyException.class)
+  public void testfindByExternalIdWithNullSearch() {
+    SessionFactory sessionFactory = mock(SessionFactory.class);
+    CountyDao countyDao = new CountyDao(sessionFactory);
+    countyDao.findByExternalId(null);
+  }
+
+  @Test
+  public void testFindByExternalIdWithId() {
```
Comment (LeonidMarushevskyi) [inline src/test/java/gov/ca/cwds/cans/dao/CountyDaoTest.java]: Just some observation regarding bottom-up TDD approach. 

Method findByExternalId has just one line, composed from the invocation of standard Hibernate API methods. The test looks more complex than actual code. Test coupled with implementation so refactoring of the source code will cause the test to fail even if the actual functionality works the same. For example, we will decide to use some custom query to retrieve information from DB instead of Natural Id. The test will break because it depends on Natural Id methods which is details of implementation.

Specific of our REST API applications is that most of the code is a simple movement of data from one layer to another. Unit tests of such code usually represent a bunch of complex mock setup and test which tests very basic logic. That is why we are using top-down TDD approach where we are starting from a functional in-memory test using fixtures to define REST contract and then moving down to the implementation of layers like resource, service, DAO, etc. We are using unit tests for complex logic where it benefits the developer and gives the ability to test complex logic and corner cases. 

Tests with extensive use of mocks not only require more efforts to develop but in long-term require more efforts to support, also limit refactoring possibilities.

File: src/test/java/gov/ca/cwds/cans/dao/PersonDaoTest.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.verify;
+
+public class PersonDaoTest {
+
+  @Test(expected = NullOrEmptyException.class)
+  public void testSearchWithNullSearch() {
+    SessionFactory sessionFactory = mock(SessionFactory.class);
+    PersonDao personDao = new PersonDao(sessionFactory);
+    personDao.search(null, "");
```
Comment (LeonidMarushevskyi) [inline src/test/java/gov/ca/cwds/cans/dao/PersonDaoTest.java]: Need a test for CountyId null.

File: src/main/java/gov/ca/cwds/cans/dao/PersonDao.java [hunk 3]
```diff
@@ -21,11 +23,12 @@ public PersonDao(@CansSessionFactory final SessionFactory sessionFactory) {
     super(sessionFactory);
   }
 
-  public Collection<Person> search(SearchPersonPo searchPo) {
+  public Collection<Person> search(SearchPersonPo searchPo, String countyId) {
```
[the remark this answers is not in the document]
Comment (edgibbs) [reply src/main/java/gov/ca/cwds/cans/dao/PersonDao.java]: The reason for the extra param is because countyId is supplied not from search through the API, but by fetching it from the logged in user's county.  It could be added as a param to `SearchPersonPo` (by the way what is a Po?  Normally it stands for Purchase Order?), but then it differs by a single field from SearchPersonRequest so it's not clear why it's in one, but not the other.  So my reasoning goes by adding countyId as an explicit argument it calls out that it's different then the parameters coming in through the API.

[the remark this answers is not in the document]
Comment (edgibbs) [reply src/main/java/gov/ca/cwds/cans/dao/PersonDao.java]: Added a quick check with a nicer exception.  My question, do we care if somehow we get a logged in user without a valid county and this causes an error?

File: src/main/java/gov/ca/cwds/cans/dao/PersonDao.java [hunk 4]
```diff
@@ -21,11 +23,12 @@ public PersonDao(@CansSessionFactory final SessionFactory sessionFactory) {
...
     Require.requireNotNullAndNotEmpty(searchPo);
 
     final Session session = grabSession();
     final PersonRole personRole = searchPo.getPersonRole();
+    session.enableFilter(Person.FILTER_COUNTY).setParameter(Person.PARAM_COUNTY_ID, new BigInteger(countyId));
```
[the remark this answers is not in the document]
Comment (edgibbs) [reply src/main/java/gov/ca/cwds/cans/dao/PersonDao.java]: Simple check added.

File: src/test/java/gov/ca/cwds/cans/dao/PersonDaoTest.java (hunk 2 above)
Comment (edgibbs) [reply src/test/java/gov/ca/cwds/cans/dao/PersonDaoTest.java]: Added.

File: src/test/java/gov/ca/cwds/cans/dao/CountyDaoTest.java (hunk 1 above)
Comment (edgibbs) [reply src/test/java/gov/ca/cwds/cans/dao/CountyDaoTest.java]: TDD/BDD is a design technique where you have to add a failing test before you add a line of production code.  The bulk of these are going to be unit tests that drive the design.  Framework code sometimes makes this hard and you have to make more use of mocks, but this often leads to a better design forcing a developer to make things more testable, and thus more loosely coupled.  The GOOS book (http://www.growing-object-oriented-software.com) is probably the most commonly cited explanation of this approach.

The vast bulk of tests in cans-api are high-level functional tests.  Almost all of them pass along an actual REST URL in their invocation.  I was expecting some test harnesses for the DAO layer to make some lightweight integration tests with an in-memory database.  Something like creating a single county and then finding it.  The current tests are at a much higher level and involve many concerns and responsibilities I don't want to deal with in a unit test.  They have many fixture dependencies and utility test infrastructure which make it harder to understand what is going on in the tests.

So this is probably better served by an offline discussion.

File: src/main/java/gov/ca/cwds/cans/dao/PersonDao.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (LeonidMarushevskyi) [reply src/main/java/gov/ca/cwds/cans/dao/PersonDao.java]: Generally, the user should have a county in PerryAccount. I think just parameter check is enough. It should be clear from stack trace what happened.

File: src/main/java/gov/ca/cwds/cans/service/PerryService.java [hunk 5]
```diff
@@ -1,46 +1,54 @@
...
 
-  private Person buildNewUser(PerryAccount perryAccount, String userUniqueId) {
+  private Person buildNewUser(PerryAccount perryAccount, String userUniqueId, County county) {
     return new Person()
         .setExternalId(userUniqueId)
         .setFirstName(perryAccount.getFirstName())
         .setLastName(perryAccount.getLastName())
-        .setPersonRole(PersonRole.USER);
+        .setPersonRole(PersonRole.USER)
+        .setCounty(county);
```
Comment (denys-davydov) [inline src/main/java/gov/ca/cwds/cans/service/PerryService.java]: If in `PersonService#search()` you are using county from security principal and not from this person, then you don't need to store the county with user-person here.

File: src/main/java/gov/ca/cwds/cans/domain/entity/Person.java [hunk 6]
```diff
@@ -49,15 +53,22 @@
   name = FILTER_EXTERNAL_ID,
   parameters = @ParamDef(name = PARAM_EXTERNAL_ID, type = "string")
 )
+@FilterDef(
+  name = FILTER_COUNTY,
+  parameters = @ParamDef(name = PARAM_COUNTY_ID, type = "big_integer")
+)
 @Filter(name = FILTER_PERSON_ROLE, condition = "person_role = :" + PARAM_PERSON_ROLE)
 @Filter(name = FILTER_EXTERNAL_ID, condition = "external_id = :" + PARAM_EXTERNAL_ID)
+@Filter(name = FILTER_COUNTY, condition = "county_id = :" + PARAM_COUNTY_ID)
```
Comment (denys-davydov) [inline src/main/java/gov/ca/cwds/cans/domain/entity/Person.java]: I think it would be better to join county here and to filter by county's external_id rather than primary key. This will help us to avoid additional db call to fetch county by external id prior to this filter search invocation.

File: src/test/java/gov/ca/cwds/cans/service/PerryServiceTest.java [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+import gov.ca.cwds.cans.domain.entity.Person;
+import gov.ca.cwds.cans.domain.enumeration.PersonRole;
+import gov.ca.cwds.security.realm.PerryAccount;
+import gov.ca.cwds.security.utils.PrincipalUtils;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.powermock.core.classloader.annotations.PrepareForTest;
+import org.powermock.modules.junit4.PowerMockRunner;
+
+import static org.mockito.Mockito.*;
```
Comment (denys-davydov) [inline src/test/java/gov/ca/cwds/cans/service/PerryServiceTest.java]: We do not use wildcards in imports

File: src/test/java/gov/ca/cwds/cans/service/PerryServiceTest.java [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+import static org.powermock.api.mockito.PowerMockito.mockStatic;
+
+@RunWith(PowerMockRunner.class)
+@PrepareForTest(gov.ca.cwds.security.utils.PrincipalUtils.class)
+public class PerryServiceTest {
+
+  @Test
+  public void testGetOrPersistAndGetCurrentUserSetsCounty() throws Exception {
```
Comment (denys-davydov) [inline src/test/java/gov/ca/cwds/cans/service/PerryServiceTest.java]: As we discussed earlier we are using **MethodName_ExpectedBehavior_StateUnderTest** test methods naming convention

File: src/test/java/gov/ca/cwds/cans/service/PersonServiceTest.java [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+import static org.powermock.api.mockito.PowerMockito.mockStatic;
+
+@RunWith(PowerMockRunner.class)
+@PrepareForTest(gov.ca.cwds.security.utils.PrincipalUtils.class)
+public class PersonServiceTest {
+
+  @Test
+  public void testSearchReturnsPeople() throws Exception {
```
Comment (denys-davydov) [inline src/test/java/gov/ca/cwds/cans/service/PersonServiceTest.java]: Please rename the method to comply with test methods naming convention

File: src/test/java/gov/ca/cwds/cans/dao/CountyDaoTest.java (hunk 1 above)
Comment (denys-davydov) [reply src/test/java/gov/ca/cwds/cans/dao/CountyDaoTest.java]: I agree on getting this offline, let's have a discussion on it.

File: src/main/java/gov/ca/cwds/cans/dao/PersonDao.java [hunk 10]
```diff
@@ -21,11 +23,13 @@ public PersonDao(@CansSessionFactory final SessionFactory sessionFactory) {
     super(sessionFactory);
   }
 
-  public Collection<Person> search(SearchPersonPo searchPo) {
+  public Collection<Person> search(SearchPersonPo searchPo, String countyId) {
     Require.requireNotNullAndNotEmpty(searchPo);
+    Require.requireNotNullAndNotEmpty(countyId);
```
[the remark this answers is not in the document]
Comment (edgibbs) [reply src/main/java/gov/ca/cwds/cans/dao/PersonDao.java]: So the story states that the user should only see their county's children.  There is no option to see all children, so why would this be an optional filter?

[the remark this answers is not in the document]
Comment (denys-davydov) [reply src/main/java/gov/ca/cwds/cans/dao/PersonDao.java]: DAO layer should be more universal because it is multipurpose (PersonDao will be definitely used not only by PersonService and not only for the business logic of your user story). So it has to have a good interface to fit our business needs, but it shouldn't constrain us with a business logic of a single user story. Look at it as a low-level instrument.

File: src/main/java/gov/ca/cwds/cans/domain/entity/Person.java [hunk 11]
```diff
@@ -30,28 +26,45 @@
...
 @FilterDef(
-  name = FILTER_EXTERNAL_ID,
-  parameters = @ParamDef(name = PARAM_EXTERNAL_ID, type = "string")
+    name = FILTER_COUNTY,
+    parameters = @ParamDef(name = PARAM_USERS_COUNTY_EXTERNAL_ID, type = "string")
 )
+@Filter(name = FILTER_PERSON_ROLE, condition = "person_role = :" + PARAM_PERSON_ROLE)
+@Filter(name = FILTER_EXTERNAL_ID, condition = "external_id = :" + PARAM_EXTERNAL_ID)
+@Filter(name = FILTER_COUNTY,
+    condition = "county_id IN (SELECT county.id FROM cans.county WHERE county.external_id = :" + PARAM_USERS_COUNTY_EXTERNAL_ID + ")")
```
Comment (edgibbs) [inline src/main/java/gov/ca/cwds/cans/domain/entity/Person.java]: Tried {h-schema} here as a hint, but looks like it doesn't operate within filter conditions.  Open to a quick refactoring.  Hibernate is a beast (🐘) of an ORM at this point.

**Comment by edgibbs** [comment]
retest this please

File: src/main/java/gov/ca/cwds/cans/domain/search/SearchParameters.java [hunk 12]
```diff
@@ -1,4 +1,4 @@
 package gov.ca.cwds.cans.domain.search;
 
 /** @author denys.davydov */
-public interface SearchPo {}
+public interface SearchParameters {}
```
Comment (denys-davydov) [inline src/main/java/gov/ca/cwds/cans/domain/search/SearchParameters.java]: While `Po` stands for `Parameter Object` I can't find it reasonable to rename it to `Parameters`.

File: src/main/java/gov/ca/cwds/cans/dao/PersonDao.java [hunk 13]
```diff
@@ -31,16 +34,17 @@ public Person find(Serializable primaryKey) {
...
 
-  public Collection<Person> search(SearchPersonPo searchPo) {
-    Require.requireNotNullAndNotEmpty(searchPo);
+  public Collection<Person> search(SearchPersonParameters searchPersonParameters) {
+    Require.requireNotNullAndNotEmpty(searchPersonParameters);
     authorize();
     final Session session = grabSession();
-    final PersonRole personRole = searchPo.getPersonRole();
+    final PersonRole personRole = searchPersonParameters.getPersonRole();
+    session.enableFilter(Person.FILTER_COUNTY).setParameter("usersCountyExternalId", searchPersonParameters.getUsersCountyExternalId());
```
Comment (denys-davydov) [inline src/main/java/gov/ca/cwds/cans/dao/PersonDao.java]: Consider to check `searchPersonParameters.getUsersCountyExternalId()` on not being null or empty value in **if statement** prior to the filter enabling

File: src/main/java/gov/ca/cwds/cans/domain/entity/Person.java (hunk 11 above)
Comment (denys-davydov) [reply src/main/java/gov/ca/cwds/cans/domain/entity/Person.java]: Yes, and this filter looks a little confusing. 
Did you try it without `cans.` part in internal select?

File: src/main/java/gov/ca/cwds/cans/service/PerryService.java [hunk 14]
```diff
@@ -3,26 +3,28 @@
...
 
 /** @author denys.davydov */
 public class PerryService {
 
   private final PersonDao personDao;
 
   @Inject
-  public PerryService(PersonDao personDao) {
+  public PerryService(PersonDao personDao)
+  {
```
Comment (denys-davydov) [inline src/main/java/gov/ca/cwds/cans/service/PerryService.java]: Ehm, a curly bracket on a new line, surprised :)

File: src/main/java/gov/ca/cwds/cans/service/PerryService.java [hunk 15]
```diff
@@ -3,26 +3,28 @@
...
     this.personDao = personDao;
   }
 
   public Person getOrPersistAndGetCurrentUser() {
     final PerryAccount perryAccount = PrincipalUtils.getPrincipal();
     final String userUniqueId = perryAccount.getUser();
-    final Collection<Person> users = findUserById(userUniqueId);
+    final Collection<Person> users = findUserById(userUniqueId, perryAccount.getCountyCwsCode());
```
Comment (denys-davydov) [inline src/main/java/gov/ca/cwds/cans/service/PerryService.java]: We still expect that `userUniqueId` is unique, so there is no need to add a county to this findUserById search, the outcome will be the same.

File: src/main/java/gov/ca/cwds/cans/service/PersonService.java [hunk 16]
```diff
@@ -36,8 +39,10 @@ public PersonService(
     return dao.findAll();
   }
 
-  public Collection<Person> search(final SearchPersonPo searchPo) {
-    return ((PersonDao) dao).search(searchPo);
+  public Collection<Person> search(final SearchPersonParameters searchPersonParameters) {
+    final PerryAccount perryAccount = PrincipalUtils.getPrincipal();
+    searchPersonParameters.setUsersCountyExternalId(perryAccount.getCountyCwsCode());
```
Comment (denys-davydov) [inline src/main/java/gov/ca/cwds/cans/service/PersonService.java]: 👍

File: src/main/java/gov/ca/cwds/cans/domain/search/SearchParameters.java (hunk 12 above)
Comment (denys-davydov) [reply src/main/java/gov/ca/cwds/cans/domain/search/SearchParameters.java]: And I don't mean to rename it back now ;)

File: src/main/java/gov/ca/cwds/cans/domain/entity/Person.java (hunk 11 above)
Comment (edgibbs) [reply src/main/java/gov/ca/cwds/cans/domain/entity/Person.java]: Yes, it bombs, it doesn't assume it's in cans so the syntax fails.  I tried out HQL style and {h-schema} without much luck.  I could make the schema a parameter to the filter and they pass it in from the DAO after looking up the default schema from the Hibernate session, but that seems clunky to.

`¯\_(ツ)_/¯`

Comment (edgibbs) [reply src/main/java/gov/ca/cwds/cans/domain/entity/Person.java]: Oh, and the in memory DB doesn't need the cans. prefix, but postgres apparently does.

File: src/main/java/gov/ca/cwds/cans/service/PerryService.java (hunk 14 above)
Comment (edgibbs) [reply src/main/java/gov/ca/cwds/cans/service/PerryService.java]: seems wrong, good catch

Comment (edgibbs) [reply src/main/java/gov/ca/cwds/cans/service/PerryService.java]: I'm guessing this was something with the massive merge conflict, too bad Sonar doesn't catch it, will fix.

**Comment by edgibbs** [comment]
Ugg more conflicts, I'll deal with early in the morning before jury duty.

File: src/main/java/gov/ca/cwds/cans/dao/PersonDao.java (hunk 13 above)
Comment (edgibbs) [reply src/main/java/gov/ca/cwds/cans/dao/PersonDao.java]: The requirement is that all children are filtered by counties, so it can't be an optional check.

File: src/main/java/gov/ca/cwds/cans/service/PerryService.java (hunk 15 above)
Comment (edgibbs) [reply src/main/java/gov/ca/cwds/cans/service/PerryService.java]: The search Interface for PersonDao#search requires a searchPersonParameters and within that it must not have a usersCountyExternalId.  It seems to make more sense to add a new search interface if you want a more specific search for PersonDao.  Something like PersonDao#findByUniqueUserId.  Is that what you want?

File: src/main/java/gov/ca/cwds/cans/dao/PersonDao.java (hunk 13 above)
Comment (edgibbs) [reply src/main/java/gov/ca/cwds/cans/dao/PersonDao.java]: Added an explicit null check.
